libafl/src/stages/pruning.rs

```diff
@@ -0,0 +1,150 @@
+//! Corpus pruning stage
+
+use core::marker::PhantomData;
+
+use libafl_bolts::{rands::Rand, Error};
+
+#[cfg(feature = "std")]
+use crate::events::EventRestarter;
+use crate::{
+    corpus::{Corpus, CorpusId},
+    stages::Stage,
+    state::{HasCorpus, HasRand, UsesState},
+};
+
+#[derive(Debug)]
+/// The stage to probablistically disable a corpus entry.
+/// This stage should be wrapped in a if stage and run only when the fuzzer perform restarting
+/// The idea comes from `https://mschloegel.me/paper/schiller2023fuzzerrestarts.pdf`
+pub struct CorpusPruning<EM> {
+    /// The chance of retaining this corpus
+    prob: f64,
+    phantom: PhantomData<EM>,
+}
+
+impl<EM> CorpusPruning<EM> {
+    fn new(prob: f64) -> Self {
+        Self {
+            prob,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<EM> Default for CorpusPruning<EM> {
+    fn default() -> Self {
+        Self::new(0.05)
+    }
+}
+
+impl<EM> UsesState for CorpusPruning<EM>
+where
+    EM: UsesState,
+{
+    type State = EM::State;
+}
+
+impl<E, EM, Z> Stage<E, EM, Z> for CorpusPruning<EM>
+where
+    EM: UsesState,
+    E: UsesState<State = Self::State>,
+    Z: UsesState<State = Self::State>,
+    Self::State: HasCorpus + HasRand,
+{
+    #[allow(clippy::cast_precision_loss)]
+    fn perform(
+        &mut self,
+        _fuzzer: &mut Z,
+        _executor: &mut E,
+        state: &mut Self::State,
+        _manager: &mut EM,
+    ) -> Result<(), Error> {
+        // Iterate over every corpus entry
+        let n_corpus = state.corpus().count_all();
+        let mut do_retain = vec![];
+        for _ in 0..n_corpus {
+            let r = state.rand_mut().below(100) as f64;
+            do_retain.push((self.prob * 100_f64) < r);
+        }
+
+        let corpus = state.corpus_mut();
+        for (idx, retain) in do_retain.iter().enumerate().take(n_corpus) {
+            if !retain {
+                let removed = corpus.remove(CorpusId(idx))?;
```

Comment:
Creating a raw `CorpusId` will likely fail at some point, after removing the corpus Ids will be sparse

Reply:
you need to `corpus.nth()` if you want to map an int to CorpusId

Reply:
i changed it, is it correct now?